src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -0,0 +1,157 @@
+
+                    if (written > 0)
+                    {
+                        break;
+                    }
+
+                    ExpandBuffer(ref result);
+                }
+
+                if (resultLength != 0)
+                {
+                    result[resultLength] = _separator;
+                    resultLength += 1;
+                }
+
+                resultLength += written;
+            }
+
+            int first = 0;
+            ReadOnlySpan<char> chars = name.AsSpan();
+            CharCategory previousCategory = CharCategory.Boundary;
+
+            for (int index = 0; index < chars.Length; index++)
+            {
+                char current = chars[index];
+                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+
+                if (currentCategoryUnicode == UnicodeCategory.SpaceSeparator ||
+                    currentCategoryUnicode >= UnicodeCategory.ConnectorPunctuation &&
+                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
+                {
+                    WriteWord(chars.Slice(first, index - first), ref result);
+
+                    previousCategory = CharCategory.Boundary;
+                    first = index + 1;
+
+                    continue;
+                }
+
+                if (index + 1 < chars.Length)
+                {
+                    char next = chars[index + 1];
+                    CharCategory currentCategory = currentCategoryUnicode switch
+                    {
+                        UnicodeCategory.LowercaseLetter => CharCategory.Lowercase,
+                        UnicodeCategory.UppercaseLetter => CharCategory.Uppercase,
+                        _ => previousCategory
+                    };
+
+                    if (currentCategory == CharCategory.Lowercase && char.IsUpper(next) ||
+                        next == '_')
+                    {
+                        WriteWord(chars.Slice(first, index - first + 1), ref result);
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index + 1;
+
+                        continue;
+                    }
+
+                    if (previousCategory == CharCategory.Uppercase &&
+                        currentCategoryUnicode == UnicodeCategory.UppercaseLetter &&
+                        char.IsLower(next))
+                    {
+                        WriteWord(chars.Slice(first, index - first), ref result);
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index;
+
+                        continue;
+                    }
+
+                    previousCategory = currentCategory;
+                }
+            }
+
+            WriteWord(chars.Slice(first), ref result);
+
+            name = result.Slice(0, resultLength).ToString();
+
+            if (buffer is not null)
+            {
+                ArrayPool<char>.Shared.Return(buffer, clearArray: true);
```

Comment:
Similarly, we could clear the span only instead of the entire rented buffer here.